Sources/NIOSSL/SSLCallbacks.swift

```diff
@@ -173,6 +173,84 @@ public struct PSKClientIdentityResponse: Sendable {
     }
 }
 
+public struct NIOSSLClientExtensionValues: Hashable {
+    public var serverHostname: String?
+
+    public init(serverHostname: String?) {
+        self.serverHostname = serverHostname
+    }
+}
+
+/// A callback that can used to support multiple or dynamic TLS hosts.
+///
+/// When set, this callback will be invoked once per TLS hello. The provided `NIOSSLClientExtensionValues` will contain the
+/// host name indicated in the TLS client hello, while the provided `NIOSSLContext` will be the current
+/// server context being used for the handshake.
+///
+/// Within this callback, the user can create and return a new server `NIOSSLContext` for the given host,
+/// return the same context provided if it is sufficient to complete the handshake, or return `throw` in the event
+/// of an exception.
+///
+public typealias NIOSSLContextCallback = @Sendable (NIOSSLClientExtensionValues, NIOSSLContext) -> EventLoopPromise<NIOSSLContext>
+
+/// A struct that provides helpers for working with a NIOSSLContextCallback.
+internal struct CustomContextManager {
+    private let callback: NIOSSLContextCallback
+
+    internal private(set) var state: State
+
+    init(callback: @escaping NIOSSLContextCallback) {
+        self.callback = callback
+        self.state = .notStarted
+    }
+}
+
+extension CustomContextManager {
+    internal enum State {
```

Comment:
I think it's generally nicer if this is private to `CustomContextManager`, and `loadContext` becomes responsible for correctly managing the state. It makes it much easier to confirm that the API to this type is sound.

I do recommend having a little computed property that tells the outer code whether it needs to construct a `NIOSSLClientExtensionValues`, but that should be as much of the internal state as this method shows. In fact, we could cache the values, and have that computed property be `var extensionValues: NIOSSLClientExtensionValues?`. If there are no cached values we'll build them, otherwise we just use the cached one.